src/Storages/ObjectStorage/StorageObjectStorageSource.cpp

```diff
@@ -937,9 +940,16 @@ StorageObjectStorageSource::ReaderHolder::operator=(ReaderHolder && other) noexc
 }
 
 StorageObjectStorageSource::ReadTaskIterator::ReadTaskIterator(
-    const ReadTaskCallback & callback_, size_t max_threads_count)
-    : callback(callback_)
+    const ReadTaskCallback & callback_, size_t max_threads_count, ContextPtr context_)
+    : WithContext(context_)
+    , callback(callback_)
 {
+    if (getContext()->isPreShutdownCalled())
```

Comment:
Is it true that operations prohibited in PreShutdown phase (e.g getting new tasks, are allowed in Shutdown phase?
If yes, is it correct?

Reply:
Then what is the purpose of shutdown_called flag?

From the first glance I would expect that all checks that are true for preshutdown_called should be true for shutdown_called as well.

Reply:
Ah, understood, flag set in destructor.
Yes, make sense to set preshutdown there too.